### DISCUSSION TITLE
HTTPS not working in a more complex context

### DISCUSSION THREAD

**Discussion by Kasperdelasopa** [esp8266/Arduino#8933] [General]
Hello,

i want to use the following code in a standalone test environment:

  ```
HTTPClient sender;
  WiFiClientSecure wifiClient;
  wifiClient.setInsecure();

  if (sender.begin(wifiClient, "https://timeapi.io/api/Time/current/zone?timeZone=Europe/Berlin")) {
    int httpCode = sender.GET(); 
    if (httpCode > 0) {
      if (httpCode == 200) {
        String payload = sender.getString();
        Serial.println(payload);
      }else{
        Serial.print("HTTP-Error: " +  String(httpCode));
      }
    }
    else{
      Serial.print("HTTP-Error: " +  String(httpCode));
    }
    sender.end();
  }
```

The code works fine as standalone.

When i want of use the same code in my project, i always receive the httpCode = -1

I have no idea what can cause this in my case.

As i cant provide the entire project (over 20.000 lines of code), i need your ideas, what could cause my code not to connect to the https-site.

Please provide me with some ideas! I'm running out of them :-(

**Comment by Kasperdelasopa** [comment]
Hello,

is there no body who can give me a hint, or has an idea?!

Can there the an interference with other instances of an HTTPClient for example?

I'm sung an  HTTPClient  and WifiClient in other classes as well.

**Comment by earlephilhower** [comment]
You deleted the template and aren't providing an MCVE so it's not possible to help you much.

My best guess is you are out of memory.  HTTPS needs one 17KB  and one 7KB contiguous chunk of RAM or it cannot work.  You might be able to move the `WiFiClientSecure` to a global, to have it allocate SSL memory at the start of the app, to get things working if you do have ~24KB free just not in contiguous chunks.  You can also look at MFLN which can reduce the 17K to 4K in many cases.

**Reply by Kasperdelasopa** [reply]
Thank you very much for your input!!

That was very usefull Information. One major difference of the "complex context" is that i have a higher usage of heap. 

I'm gonna try your ideas about a global client object and the MFLN  usage.

Thx!!!

**Comment by InnuendoPi** [comment]
I have a very similar problem. Example code ist working, but in a larger project it fails. Most read answer is heap or OOM. The problem occurs with sender.GET(). If you request only a few bytes, get() returns 200 and snippet works. Also downloads in chunks are working. But if you request a larger file get() returns not connected regardless of setInsecure and/or setBufferSizes or MMU settings. So snippet from Kasperdelasopa always ends with httpCode -1 or -5
Some debug outputs:
###
* SYSINFO: free heap at boot: 27392
* SYSINFO: free heap after BearSSL::WiFiClientSecure clientup: 20944
* get() a.min.js getSize: 421343 received size: 0
* get() b.min.css getSize: 171218 received size: 0
* get() c.ico getSize: 1150 received size: 1150
* get() d.ttf getSize: 8296 received size: 0
###
icon file c.ico with 1150b successful transfered in chunks of 128. all other failed in same loop. free heap 27392 does not fit ~28k (magic BearSSL limes). Maybe PR #9005 (readStringUntil) points in a new direction to solve this problem.
